torch/testing/_internal/opinfo_helper.py

```diff
@@ -0,0 +1,110 @@
+from collections import defaultdict
+import warnings
+from functools import partial
+
+import torch
+from torch.testing._internal.common_cuda import (TEST_CUDA)
+from torch.testing._internal.common_methods_invocations import (op_db, _DYNAMIC_DTYPES)
+from torch.testing import (all_types_and_complex_and,
+                           all_types_and_complex,
+                           # all_types_and_half, # not exported by testing at the moment
+                           all_types,
+                           complex_types,
+                           floating_and_complex_types,
+                           # floating_types_and_half, # not exported by testing at the moment
+                           floating_types,
+                           integral_types,
+                           )
+
+DTYPE_DISPATCH = (
+    floating_types,
+    floating_and_complex_types,
+    integral_types,
+    all_types,
+    complex_types,
+    all_types_and_complex,
+)
+
+# Better way to acquire devices?
+DEVICES = ['cpu'] + (['cuda'] if TEST_CUDA else [])
+
+if not TEST_CUDA:
+    warnings.warn("WARNING: CUDA is not available, information pertaining to CUDA could be wrong")
+
+
+class OpInfoHelper:
+    def __init__(self, op_info):
+        self.op = op_info
+        self.sample_inputs_fn = op_info.sample_inputs_func
+
+    def get_supported_dtypes(self):
+        # Return a map of (device, supported_dtypes)
+        device_dtype = defaultdict(list)
+        for device in DEVICES:
+            for dtype in all_types_and_complex_and(torch.bool, torch.bfloat16, torch.half):
+                sample = self.sample_inputs_fn(self.op, device, dtype, False)[0]
+                try:
+                    self.op(sample.input, *sample.args, **sample.kwargs)
+                    device_dtype[device].append(dtype)
+                except RuntimeError as re:
+                    if "not implemented for" not in str(re):
+                        raise re
+                    pass
+
+        return device_dtype
+
+    def has_out_support(self, device, dtype):
+        try:
+            sample = self.sample_inputs_fn(None, device, dtype, False)[0]
+            op_out = partial(self.op, sample.input, *sample.args, **sample.kwargs)
+            op_out(out=torch.empty_like(sample.input))
+            return True
+        except TypeError as te:
+            if "unexpected keyword argument 'out'" not in str(te):
+                raise te
+            return False
+
+    def dtype_dispatch_hint(self, dtypes):
+        # Find the closest dispatcher from the set of available dispatchers
+
+        # Metric for finding the closest dispatcher
+        # Intersection Over Union
+        def iou(dispatch_dtypes):
```

Comment:
I think this can be simplified a bit:

- check for an exact match by going through functions without `_AND`
- if that fails, find the `_AND` function with the greatest number of matches and no mismatches
  - `_AND` in the set difference

Reply:
Will the current version, for example, incorrectly suggest trying to extend functions without `_AND`?

Reply:
> Will the current version, for example, incorrectly suggest trying to extend functions without _AND?

Not entirely sure if I understand the question correctly, but here it goes

For `topk` which supports dtypes satisfied by `all_types` dispatch.

Below is the generated output, hint is `# hint: all_types + ()`
```
OpInfo(topk,
       # hint: all_types + (),
       dtypes=[torch.float32, torch.float64, torch.uint8, torch.int8, torch.int16, torch.int32, torch.int64],
       sample_inputs_func=sample_inputs_topk)
```